src/ast/analysis/TypeConstraints.cpp

```diff
@@ -0,0 +1,635 @@
+            };
+
+            overloads = filterNot(std::move(overloads), [&](const IntrinsicFunctorInfo& x) -> bool {
+                if (!x.variadic && args.size() != x.params.size()) return true;  // arity mismatch?
+
+                for (size_t i = 0; i < args.size(); ++i)
+                    if (!possible(x.params[x.variadic ? 0 : i], args[i])) return true;
+
+                return !possible(x.result, result);
+            });
+
+            bool changed = false;
+            auto newResult = [&]() -> std::optional<TypeSet> {
+                if (0 == overloads.size()) return TypeSet();
+                if (1 < overloads.size()) return {};
+
+                auto& overload = overloads.front().get();
+                // `ord` is freakin' magical: it has the signature `a -> Int`.
+                // As a consequence, we might be given non-primitive arguments (i.e. types for which
+                // `TypeEnv::getConstantType` is undefined).
+                // Handle this by not imposing constraints on the arguments.
+                if (overload.op != FunctorOp::ORD) {
+                    for (size_t i = 0; i < args.size(); ++i) {
+                        auto argTy = overload.params[overload.variadic ? 0 : i];
+                        auto& currArg = assigment[args[i]];
+                        auto newArg = subtypesOf(currArg, argTy);
+                        changed |= currArg != newArg;
+                        // 2020-05-09: CI linter says to remove `std::move`, but clang-tidy-10 is happy.
+                        currArg = std::move(newArg);  // NOLINT
+                    }
+                }
+
+                if (nonMonotonicUpdate || subtypeResult) {
+                    return subtypesOf(assigment[result], overload.result);
+                } else {
+                    nonMonotonicUpdate = true;
+                    return TypeSet{typeEnv.getConstantType(overload.result)};
+                }
+            }();
+
+            if (newResult) {
+                auto& curr = assigment[result];
+                changed |= curr != *newResult;
+                // 2020-05-09: CI linter says to remove `std::move`, but clang-tidy-10 is happy.
+                curr = std::move(*newResult);  // NOLINT
+            }
+
+            return changed;
+        }
+
+        void print(std::ostream& out) const override {
+            // TODO (darth_tytus): is this description correct?
+            out << "∃ t : " << result << " <: t where t is a base type";
+        }
+    };
+
+    return std::make_shared<C>(
+            typeEnv, std::move(overloads), std::move(result), std::move(args), subtypeResult);
+}
+
+/**
+ * Constraint on record type and its elements.
+ */
+static TypeConstraint isSubtypeOfComponent(
+        const TypeVar& elementVariable, const TypeVar& recordVariable, size_t index) {
+    struct C : public Constraint<TypeVar> {
+        TypeVar elementVariable;
+        TypeVar recordVariable;
+        unsigned index;
```

Comment:
This is a little bizarre - index is std::size_t, int, and unsigned. Standardise on ... size_t.